### PR TITLE
Fix nested arrays (by reworking array handling)

### LINKED ISSUE #464: Nested arrays badly mangled
To reproduce:
```
    #[test]
    fn list() {
        let jtxt = r#" { "proxy_ports": [ [1,2] ] } "#;
        let jval: serde_json::Value = serde_json::from_str(jtxt).unwrap();
        //dbg!(&jval);
        let jcfg = config::Config::try_from(&jval).unwrap();
        //dbg!(&jcfg);
        let jret: serde_json::Value = jcfg.clone().try_deserialize().unwrap();
        println!("json roundtrip: {jret:#?}");

        let ttxt = r#" proxy_ports = [ [1,2] ] "#;
        let tval: toml::Value = toml::from_str(ttxt).unwrap();
        //dbg!(&tval);
        let tcfg = config::Config::try_from(&tval).unwrap();
        //dbg!(&tcfg);
        let tret: Result<String, _> = tcfg.clone().try_deserialize()
            .map(|v: toml::Value| toml::to_string(&v).unwrap());
        println!("toml roundtrip: {tret:#?}");
    }
```

Expected output:
```
json roundtrip: Object {
    "proxy_ports": Array [
        Array [
            Number(1),
            Number(2),
        ],
    ],
}
toml roundtrip: Ok(
    "proxy_ports = [[1, 2]]\n"
)
```

Actual output:

```
json roundtrip: Object {
    "proxy_ports": Array [
        Null,
        Number(1),
        Number(2),
    ],
}
toml roundtrip: Err(
    invalid type: unit value, expected any valid TOML value,
)
```

Uncommenting the dbgs in the repro above shows this, which seems entirely wrong:

```
[crates/arti/src/onion_proxy.rs:143] &jcfg = Config {
    defaults: {},
    overrides: {
        Subscript(
            Identifier(
                "proxy_ports",
            ),
            1,
        ): Value {
            origin: None,
            kind: I64(
                1,
            ),
        },
        Subscript(
            Identifier(
                "proxy_ports",
            ),
            2,
        ): Value {
            origin: None,
            kind: I64(
                2,
            ),
        },
    },
    sources: [],
    cache: Value {
        origin: None,
        kind: Table(
            {
                "proxy_ports": Value {
                    origin: None,
                    kind: Array(
                        [
                            Value {
                                origin: None,
                                kind: Nil,
                            },
                            Value {
                                origin: None,
                                kind: I64(
                                    1,
                                ),
                            },
                            Value {
                                origin: None,
                                kind: I64(
                                    2,
                                ),
                            },
                        ],
                    ),
                },
            },
        ),
    },
}
```

### PR DESCRIPTION
**Pull request by ijackson**
Fixes #464.

### CODE REVIEW

**Review by polarathene — APPROVED** [review]
Not a proper review, but verified it fixes the issue 👍 

@matthiasbeyer (friendly ping) will hopefully have some time to spare to review / merge the bugfix.

File: src/ser.rs [hunk 1]
```diff
@@ -8,87 +9,83 @@ use crate::Config;
...
+
+/// Serializer for numbered sequences
+///
+/// This wrapper is present when we are outputting a sequence (numbered indices).
+/// Making this a separate type centralises the handling of sequences
+/// and ensures we don't have any call sites for `ser::SerializeSeq::serialize_element`
+/// that don't do the necessary work of `SeqSerializer::new`.
+///
+/// Existence of this wrapper implies that `.0.keys.last()` is
+/// `Some(SerKey::Key(next_index))`.
```
Comment (matthiasbeyer) [inline src/ser.rs:30]: Here the `Key` variant is used, but such a variant doesn't exist.

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Only one nitpick

File: src/ser.rs (hunk 1 above)
Comment (ijackson) [reply src/ser.rs:30]: Thanks for the review.  I'll fix this on Monday.

**Comment by matthiasbeyer** [comment]
Squashed the fixup commits for you and signed off with my own signoff because you didn't. I assume that you're alright with the signoff and its implications? If yes, I'll merge.

**Review by matthiasbeyer — APPROVED** [review]

**Comment by matthiasbeyer** [comment]
Meh. CI broke because a dependency of ours does no longer build with 1.66.0.

I'll file a PR to update MSRV, after that is merged please rebase this (and signoff your commits while you're at it! :laughing: ) please!

**Comment by ijackson** [comment]
> Squashed the fixup commits for you and signed off with my own signoff because you didn't. I assume that you're alright with the signoff and its implications. Also, if you want, feel free to add your signoff to the commits as well!

I think by signoff you mean that you want my confirmation on the statements in the [Developer Certificate of Origin](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) , v1.1 presumably?  I'm happy to confirm that all those statements apply to all of my contributions to `config-rs`, thanks.

You may want to copy `DEVELOPER-CERTIFICATE` into your tree.  Otherwise it's slightly ambiguous what `Signed-off-by` means: what is the committer signing off on?

**Comment by ijackson** [comment]
I read in https://github.com/mehcode/config-rs/pull/483#issuecomment-1775535442 that you're not intending a release any time soon.  That's awkward for us.  We need this bugfix for a feature we want to release at least a preview of fairly soon.  Our  current workaround is not really suitable.  If you don't intend to release soon with this fix, we may need to publish a forked crate to crates.io, which I would rather avoid!

**Comment by matthiasbeyer** [comment]
Do I read this correctly that _this PR_ is what you'd need for your project?
You're currently using 0.13.3 I assume?

Is there a chance that we can backport the fix to the `release-0.13.x` branch and publish 0.13.4 with the fix only?
Would that be a solution for you?

**Comment by ijackson** [comment]
Rebased, thanks.

Yes, a backport of this to the release branch would be perfect for us.  I can make an MR for that right away if you like.

**Comment by ijackson** [comment]
Not sure I understand why the MSRV CI test is still failing.  Did I rebase onto the wrong thing?

**Comment by matthiasbeyer** [comment]
> I can make an MR for that right away if you like.

Please do! :heart:

**Comment by matthiasbeyer** [comment]
> Not sure I understand why the MSRV CI test is still failing. Did I rebase onto the wrong thing?

You did not rebase on latest master...

**Comment by ijackson** [comment]
I think my push didn't take or something?  Done now I think.  Sorry.

**Comment by matthiasbeyer** [comment]
> I think my push didn't take or something? Done now I think. Sorry.

Nah don't worry! :+1:

**Pull request merged**